floats/floats.go

```diff
@@ -515,31 +515,109 @@ func NaNPayload(f float64) (payload uint64, ok bool) {
 // Nearest returns the index of the element in s
 // whose value is nearest to v.  If several such
 // elements exist, the lowest index is returned.
-// Panics if len(s) == 0.
+// Nearest panics if len(s) == 0.
 func Nearest(s []float64, v float64) int {
+	if len(s) == 0 {
+		panic("floats: zero length slice")
+	}
+	switch {
+	case math.IsNaN(v):
+		return 0
+	case math.IsInf(v, 1):
+		return MaxIdx(s)
+	case math.IsInf(v, -1):
+		return MinIdx(s)
+	}
 	var ind int
 	dist := math.Abs(v - s[0])
 	for i, val := range s {
 		newDist := math.Abs(v - val)
-		if newDist < dist {
+		// A NaN distance will not be closer.
+		if math.IsNaN(newDist) {
+			continue
+		}
+		// The inverted test here is used to
+		// ensure that a test of newDist against
+		// a NaN dist results in a change to dist
+		// and the current index value.
+		if !(newDist >= dist) {
 			dist = newDist
 			ind = i
 		}
 	}
 	return ind
 }
 
-// NearestWithinSpan return the index of a hypothetical vector created
+// NearestIdx return the index of a hypothetical vector created
 // by Span with length n and bounds l and u whose value is closest
-// to v. NearestWithinSpan panics if u < l. If the value is greater than u or
-// less than l, the function returns -1.
-func NearestWithinSpan(n int, l, u float64, v float64) int {
-	if u < l {
-		panic("floats: upper bound greater than lower bound")
+// to v. NearestIdx panics if n is zero.
+func NearestIdx(n int, l, u float64, v float64) int {
+	if n == 0 {
```

Comment:
I think we should also panic if `l > u` (and in `Span`)

Reply:
I don't see any basis for prohibiting a descending span.